packages/router/utils.ts

```diff
@@ -161,40 +161,39 @@ export interface DataRouteObject extends RouteObject {
   id: string;
 }
 
-type Star = "*";
+// Recursive helper for finding path parameters in the absence of wildcards
+type _PathParam<Path extends string> =
+  // split path into individual path segments
+  Path extends `${infer L}/${infer R}` ? _PathParam<L> | _PathParam<R> :
+  // find params after `:`
+  Path extends `${string}:${infer Param}` ? Param :
+  // otherwise, there aren't any params present
+  never
+
 /**
- * @private
- * Return string union from path string.
- * @example
- * PathParam<"/path/:a/:b"> // "a" | "b"
- * PathParam<"/path/:a/:b/*"> // "a" | "b" | "*"
+ * Examples:
+ * "/a/b/*" -> "*"
+ * ":a" -> "a"
+ * "/a/:b" -> "b"
+ * "/a/blahblahblah:b" -> "b"
+ * "/:a/:b" -> "a" | "b"
+ * "/:a/b/:c/*" -> "a" | "c" | "*"
  */
 type PathParam<Path extends string> =
-  // Check path string starts with slash and a param string.
-  Path extends `:${infer Param}/${infer Rest}`
-    ? Param | PathParam<Rest>
-    : // Check path string is a param string.
-    Path extends `:${infer Param}`
-    ? Param
-    : // Check path string ends with slash and a param string.
-    Path extends `${any}/:${infer Param}`
-    ? PathParam<`:${Param}`>
-    : // Check path string ends with slash and a star.
-    Path extends `${any}/${Star}`
-    ? Star
-    : // Check string is star.
-    Path extends Star
-    ? Star
-    : never;
```

Comment:
Note we temporarily lost the ability to have arbitrary prefixes for params (e.g. `blahblahblah:foo` has the param `foo`) in #9090 , but this PR restores that ability via `Path extends `${string}:${infer Param}` ? Param :` in `_PathParam`.

Reply:
nice - checks out on the original examples and some new ones with prefixed params: https://www.typescriptlang.org/play?#code/PTAECUFMGMFcCcDOBLAbpUALSAbADpPKAGYD2RxyAdgCbUDmoeAhgC6ZPPzMC2krhRKGqh2GZgCNEkKtAylioAO7IcNaFxqIAUKwCeBUAH0ACm0xnuPADxn2oSAA8BtIYlbwGAPlABebaCgIKCIeDjIrJz21KykwrRoyDSwzDhRHNL0fFSsOoF2HE4uWqAABgAkAN7UxISgADIAvsBVNXXgjaWgAPzGBZa81vU+AD595gM24D4AXAFBYJS0nFZCzMQCRKUzpfMFDs4yJRWV7p5U9I0zrVS1RJOdPaCToHOBwaRi8CrSADSi2Hg4iBVAA5JFmFQ9CteEI8EDpDl5lRIOh4NptMAAFRYgJY0AAUUcvDCkEQb3xACJgMxgBJsZTQABaHyUrGUvGgSkzZiMllc3mc6m0mYSPmssVCml0nDMTASWXyxWi8VcyWgKnAHlasXM1m80BjSnqzXa+kzaAMvUCxlG6C2rnsvHAXQGDD9LiDfZFI5uDzePzzYLQbDQADWwkULGiQgAVrB3KBmMpVOpNHtzAdikI2YzernXkGwDhSKQI2QiCo1Bp4DQk5ExA5lgoARho5gM-Yfa4yjc7hAyaxmliuvn2YbxuxJtYoO5ZkXQCWyyRyDCeEIRI3JIi5KAW1W07W8pOLJ6bAUvBjtMRYLJWMhSFRQPQZIQ2JACrZM92SmdvAAKeZ2xmZ5zF+ICz3JUBKnmQIAG0w0gaERA9Kwv3YLwAF0QL-C4AG55kaPxoKI5g1ihbQAEocP9C5oPmIFWAQJ9KQ5RoMRfFFuAEAp-0pXl-kqRpKIWaC4McGjzkYAAyEjsJCWj6Dw9iQAAPW6DjX24j9zD4nlKUE4TgkqcTJIYUBZKE+TcKUm0zPwlTgHUzSuPfXjqX0wzKOM0yFKkiy5Ps2z+O5PyGGUzEnI07ROLfHjdI82kxS8nyJLCujLMaazFLwuz0qUxznJirS3ISrURWSkjvLAEy0psgKrKC3KQqatVQpsiK1Oi2LtPc6UVRSmrfPqzLsqk5qxSawrupK+L2D48q6S1e1BrEurFIarLWpa-LmvtKbIqKnrSvm3NVtqoLNrG8LHXanLppcuKdNOhlzuGjbRu29kDq6x7erK2ksTpAyqtSy7Pt2h7itcubMAWpLXtBob1v8iGOtun6or+k64eFHVEaE6q1vBwLdoxyHDpmmHntxxb4UgShHCZUUCaM5GScasnc0x5ygA